### ISSUE TITLE
Segfault when Rails enters parallell testing mode

### ISSUE THREAD

**Issue by buffpojken** [rgeo/rgeo-proj4#39]
### Steps to reproduce
Include rgeo in a Rails 7 project. Add enough tests to trigger tests running in parallell (default is 50 tests). RGeo will segfault with the attached stack trace. 

### Expected behavior
Tests to run as usual

### Actual behavior
Segfault in proj4-communication.

### System configuration

**Ruby version**:
3.1.0

**OS**:
MacOS

### Stack trace

`/Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-proj4-3.1.1/lib/rgeo/coord_sys/proj4.rb:191: [BUG] Segmentation fault at 0x0000000104c38a70
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]`

```-- Control frame information -----------------------------------------------
c:0049 p:---- s:0264 e:000263 CFUNC  :_create
c:0048 p:0052 s:0258 e:000257 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-proj4-3.1.1/lib/rgeo/coord_sys/proj4.rb:191
c:0047 p:0392 s:0251 e:000250 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-2.4.0/lib/rgeo/geos/capi_factory.rb:66
c:0046 p:0097 s:0234 e:000233 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-2.4.0/lib/rgeo/geos/interface.rb:200
c:0045 p:0060 s:0228 e:000227 BLOCK  /Users/pekkaakerstrom/repos/pattern/app/models/local_plan.rb:113 [FINISH]
c:0044 p:---- s:0217 e:000216 IFUNC 
c:0043 p:---- s:0214 e:000213 CFUNC  :each
c:0042 p:---- s:0211 e:000210 CFUNC  :each_with_index
c:0041 p:0006 s:0207 e:000206 BLOCK  /Users/pekkaakerstrom/repos/pattern/app/models/local_plan.rb:107
c:0040 p:0012 s:0204 e:000203 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/transaction [FINISH]
c:0039 p:---- s:0197 e:000196 CFUNC  :handle_interrupt
c:0038 p:0029 s:0192 e:000191 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/concurrency/load_interlock_aware_monit [FINISH]
c:0037 p:---- s:0189 e:000188 CFUNC  :handle_interrupt
c:0036 p:0021 s:0184 e:000183 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/concurrency/load_interlock_aware_monit
c:0035 p:0008 s:0179 e:000178 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/transaction
c:0034 p:0053 s:0172 e:000171 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st
c:0033 p:0011 s:0163 e:000162 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/transactions.rb:209
c:0032 p:0016 s:0157 e:000156 METHOD /Users/pekkaakerstrom/repos/pattern/app/models/local_plan.rb:106
c:0031 p:0032 s:0147 e:000146 BLOCK  /Users/pekkaakerstrom/repos/pattern/test/models/local_plan_test.rb:35 [FINISH]
c:0030 p:0018 s:0142 e:000141 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:98
c:0029 p:0002 s:0139 e:000138 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:195
c:0028 p:0004 s:0134 e:000133 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:95
c:0027 p:0015 s:0131 e:000130 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:296
c:0026 p:0004 s:0126 e:000125 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:94
c:0025 p:0029 s:0123 E:001ac0 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:391
c:0024 p:0044 s:0115 E:001f78 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:243
c:0023 p:0004 s:0108 E:0014f0 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:93
c:0022 p:0008 s:0104 e:000103 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/executor/test_helper.rb:5
c:0021 p:0012 s:0101 e:000100 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/execution_wrapper.rb:105
c:0020 p:0016 s:0096 e:000095 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/executor/test_helper.rb:5
c:0019 p:0008 s:0090 e:000089 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:1059
c:0018 p:0015 s:0083 e:000082 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:50
c:0017 p:0029 s:0080 E:001440 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:391
c:0016 p:0029 s:0072 E:001238 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:378
c:0015 p:0049 s:0065 E:001338 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:49
c:0014 p:0011 s:0056 e:000055 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:38
c:0013 p:0053 s:0051 e:000050 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:27 [FINISH]
c:0012 p:---- s:0048 e:000047 CFUNC  :fork
c:0011 p:0004 s:0044 e:000043 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:15
c:0010 p:0018 s:0040 e:000039 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization.rb:37 [FINISH]
c:0009 p:---- s:0036 e:000035 IFUNC 
c:0008 p:---- s:0033 e:000032 CFUNC  :times
c:0007 p:---- s:0030 e:000029 CFUNC  :each
c:0006 p:---- s:0027 e:000026 CFUNC  :map
c:0005 p:0008 s:0023 e:000022 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization.rb:36
c:0004 p:0023 s:0019 e:000018 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelize_executor.rb:18
c:0003 p:0162 s:0015 e:000014 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:156
c:0002 p:0073 s:0008 E:001c50 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:83 [FINISH]
c:0001 p:0000 s:0003 E:001960 (none) [FINISH]```

**Comment by Pe-co** [comment]
@keithdoggett, I'm taking the ball from @buffpojken since his plate is fuller than mine atm. 
I tested again after upgrading and get the same result, 

```
-- Control frame information -----------------------------------------------
c:0066 p:---- s:0353 e:000352 CFUNC  :_create
c:0065 p:0079 s:0347 e:000346 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-proj4-4.0.0/lib/rgeo/coord_sys/proj4.rb:249
c:0064 p:0099 s:0340 e:000339 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-3.0.0/lib/rgeo/impl_helper/utils.rb:28
c:0063 p:0250 s:0333 e:000332 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-3.0.0/lib/rgeo/geos/capi_factory.rb:56
c:0062 p:0099 s:0318 e:000317 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-3.0.0/lib/rgeo/geos/interface.rb:174
c:0061 p:0038 s:0312 e:000311 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-3.0.0/lib/rgeo/cartesian/interface.rb:31
c:0060 p:0045 s:0307 e:000306 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-activerecord-7.0.1/lib/rgeo/active_record/spatial_factory_store.rb:42
c:0059 p:0016 s:0302 e:000301 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-activerecord-7.0.1/lib/rgeo/active_record/spatial_factory_store.rb:21
c:0058 p:0014 s:0297 e:000296 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/rgeo-activerecord-7.0.1/lib/rgeo/active_record/spatial_factory_store.rb:29
c:0057 p:0031 s:0292 e:000291 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-postgis-adapter-8.0.1/lib/active_record/connection_adapters/postgi
c:0056 p:0055 s:0288 e:000286 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-postgis-adapter-8.0.1/lib/active_record/connection_adapters/postgi
c:0055 p:0005 s:0282 e:000281 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-postgis-adapter-8.0.1/lib/active_record/connection_adapters/postgi
c:0054 p:0028 s:0277 e:000276 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-postgis-adapter-8.0.1/lib/active_record/connection_adapters/postgi
c:0053 p:0013 s:0272 e:000271 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-postgis-adapter-8.0.1/lib/active_record/connection_adapters/postgi
c:0052 p:0027 s:0266 e:000264 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st [FINISH]
c:0051 p:---- s:0259 e:000258 IFUNC 
c:0050 p:---- s:0256 e:000255 CFUNC  :each
c:0049 p:---- s:0253 e:000252 CFUNC  :map
c:0048 p:0076 s:0249 e:000248 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st [FINISH]
c:0047 p:---- s:0244 e:000243 CFUNC  :map
c:0046 p:0017 s:0240 e:000239 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st
c:0045 p:0016 s:0228 e:000227 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st [FINISH]
c:0044 p:---- s:0223 e:000222 IFUNC 
c:0043 p:---- s:0220 e:000219 CFUNC  :each
c:0042 p:---- s:0217 e:000216 CFUNC  :filter_map
c:0041 p:0005 s:0213 e:000212 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st
c:0040 p:0009 s:0208 e:000207 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_st
c:0039 p:0031 s:0199 E:001c68 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/fixtures.rb:630 [FINISH]
c:0038 p:---- s:0193 e:000192 CFUNC  :each
c:0037 p:0012 s:0189 E:0026c8 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/fixtures.rb:621
c:0036 p:0024 s:0182 e:000181 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/fixtures.rb:607
c:0035 p:0088 s:0172 E:000850 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/fixtures.rb:567
c:0034 p:0026 s:0160 e:000159 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/test_fixtures.rb:271
c:0033 p:0140 s:0155 E:000058 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/test_fixtures.rb:125
c:0032 p:0003 s:0150 e:000149 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activerecord-7.0.3.1/lib/active_record/test_fixtures.rb:10
c:0031 p:0004 s:0146 e:000145 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/setup_and_teardown.rb:40
c:0030 p:0003 s:0142 e:000141 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:96
c:0029 p:0002 s:0139 e:000138 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:195
c:0028 p:0004 s:0134 e:000133 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:95
c:0027 p:0015 s:0131 e:000130 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:296
c:0026 p:0004 s:0126 e:000125 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:94
c:0025 p:0029 s:0123 E:001840 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:391
c:0024 p:0044 s:0115 E:002428 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:243
c:0023 p:0004 s:0108 E:000240 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest/test.rb:93
c:0022 p:0008 s:0104 e:000103 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/executor/test_helper.rb:5
c:0021 p:0012 s:0101 e:000100 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/execution_wrapper.rb:105
c:0020 p:0016 s:0096 e:000095 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/executor/test_helper.rb:5
c:0019 p:0008 s:0090 e:000089 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:1059
c:0018 p:0015 s:0083 e:000082 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:50
c:0017 p:0029 s:0080 E:000ef0 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:391
c:0016 p:0029 s:0072 E:001398 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:378
c:0015 p:0049 s:0065 E:001b98 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:49
c:0014 p:0011 s:0056 e:000055 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:38
c:0013 p:0053 s:0051 e:000050 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:27 [FINISH]
c:0012 p:---- s:0048 e:000047 CFUNC  :fork
c:0011 p:0004 s:0044 e:000043 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization/worker.rb:15
c:0010 p:0018 s:0040 e:000039 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization.rb:37 [FINISH]
c:0009 p:---- s:0036 e:000035 IFUNC 
c:0008 p:---- s:0033 e:000032 CFUNC  :times
c:0007 p:---- s:0030 e:000029 CFUNC  :each
c:0006 p:---- s:0027 e:000026 CFUNC  :map
c:0005 p:0008 s:0023 e:000022 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelization.rb:36
c:0004 p:0023 s:0019 e:000018 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/activesupport-7.0.3.1/lib/active_support/testing/parallelize_executor.rb:18
c:0003 p:0162 s:0015 e:000014 METHOD /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:156
c:0002 p:0073 s:0008 E:0023b0 BLOCK  /Users/pekkaakerstrom/.rvm/gems/ruby-3.1.2/gems/minitest-5.16.3/lib/minitest.rb:83 [FINISH]
c:0001 p:0000 s:0003 E:001d50 (none) [FINISH]
```

**Comment by keithdoggett** [comment]
@Pe-co thanks for trying that.

I tried to replicate in proj4 with a test like this

```ruby
  def test_create_parallel
    epsg_codes = (2000..2049).to_a
    Parallel.each(epsg_codes, in_parallel: 8) do |epsg_code|
      proj = RGeo::CoordSys::Proj4.create(epsg_code)
      assert_equal("EPSG:#{epsg_code}", proj.auth_name)
    end
  end
```

but couldn't reproduce the issue.

I also created a rails app from scratch and added enough tests to trigger parallel tests, but still couldn't get a seg fault.

A few questions:
* What version of proj are you using?
* The segfault is happening when minitest is spawning child processes not during actual testing?
* If possible can you share your `SpatialFactoryStore` configuration?

**Comment by Pe-co** [comment]
This is the rgeo initializer we use
```
RGeo::ActiveRecord::SpatialFactoryStore.instance.tap do |store|
  store.register(RGeo::Geographic.spherical_factory(srid: 4326, uses_lenient_assertions: true), {sql_type: 'geography'})
end
```

**Comment by keithdoggett** [comment]
Thanks for that extra info @Pe-co and apologies about the delay, I was away the last few weeks.

I'll try to replicate with that version of Proj.

I've been thinking about this some and I am a little confused why this issue is happening with multiprocessing. Issues like this typically happen when trying multi-threading (ex. https://github.com/rgeo/rgeo/issues/307), but I'm not sure why this issue would only be happening when we have multiple processes. Maybe there's some kind of shared resource that Proj writes to that could be causing the issue.

**Comment by Pe-co** [comment]
`Maybe the way minitest spawns additional processes is causing issues.`
There seems to be some child process relation from the prints, so perhaps you are right.
```
Running 108 tests in parallel using 10 processes
DEBUGGER: Attaching after process 74415 fork to child process 74421
DEBUGGER[bin/rails#74422]: Attaching after process 74415 fork to child process 74422
DEBUGGER[bin/rails#74423]: Attaching after process 74415 fork to child process 74423
DEBUGGER[bin/rails#74424]: Attaching after process 74415 fork to child process 74424
DEBUGGER[bin/rails#74425]: Attaching after process 74415 fork to child process 74425
DEBUGGER[bin/rails#74426]: Attaching after process 74415 fork to child process 74426
DEBUGGER[bin/rails#74427]: Attaching after process 74415 fork to child process 74427
DEBUGGER[bin/rails#74428]: Attaching after process 74415 fork to child process 74428
DEBUGGER[bin/rails#74429]: Attaching after process 74415 fork to child process 74429
Run options: --seed 61645

# Running:

DEBUGGER[bin/rails#74430]: Attaching after process 74415 fork to child process 74430
```

**Comment by keithdoggett** [comment]
I tried to replicate the issue in a fresh rails app with the following setup. I'm also using proj 9.0.1.

```ruby
# db/schema.rb

ActiveRecord::Schema[7.0].define(version: 2023_04_17_124002) do
  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"
  enable_extension "postgis"

  create_table "buildings", force: :cascade do |t|
    t.string "name"
    t.geometry "boundary", limit: {:srid=>3006, :type=>"st_polygon"}
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "geo_buildings", force: :cascade do |t|
    t.string "name"
    t.geography "boundary", limit: {:srid=>4326, :type=>"st_polygon", :geographic=>true}
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

end
```

```ruby
# config/initializers/spatial_factory_store.rb
# frozen_string_literal: true

RGeo::ActiveRecord::SpatialFactoryStore.instance.tap do |config|
  config.register(RGeo::Geographic.spherical_factory(srid: 4326, uses_lenient_assertions: true),
                  { sql_type: 'geography' })
end
```

```ruby
# test/models/building_test.rb

require 'test_helper'

class BuildingTest < ActiveSupport::TestCase
  1000.times do |i|
    define_method "test_number_#{i}" do
      building = Building.new
      polygon = "POLYGON ((0 0, 0 #{i}, #{i} #{i}, #{i} 0, 0 0))"
      building.boundary = polygon
      building.name = i.to_s

      assert_equal(building.boundary.as_text, polygon)
      building.save

      building = Building.last
      assert_equal(building.boundary.as_text, polygon)

      geo_fac = RGeo::Geographic.spherical_factory(srid: 4326)
      geo_building = GeoBuilding.new

      polygon = geo_fac.parse_wkt("POLYGON ((0.0 0.0, 0.0 #{(i % 5) + 1}.0, #{(i % 5) + 1}.0 #{(i % 5) + 1}.0, #{(i % 5) + 1}.0 0.0, 0.0 0.0))")

      geo_building.boundary = polygon
      geo_building.name = i.to_s

      assert_equal(geo_building.boundary, polygon)
      geo_building.save

      geo_building = GeoBuilding.last
      assert_equal(geo_building.boundary, polygon)
    end
  end
end
```

```ruby
# Gemfile
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '3.1.2'

# Bundle edge Rails instead: gem "rails", github: "rails/rails", branch: "main"
gem 'rails', '~> 7.0.4', '>= 7.0.4.3'

# Use postgresql as the database for Active Record
gem 'pg', '~> 1.1'

# Use the Puma web server [https://github.com/puma/puma]
gem 'puma', '~> 5.0'

# Build JSON APIs with ease [https://github.com/rails/jbuilder]
# gem "jbuilder"

# Use Redis adapter to run Action Cable in production
# gem "redis", "~> 4.0"

# Use Kredis to get higher-level data types in Redis [https://github.com/rails/kredis]
# gem "kredis"

# Use Active Model has_secure_password [https://guides.rubyonrails.org/active_model_basics.html#securepassword]
# gem "bcrypt", "~> 3.1.7"

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]

# Reduces boot times through caching; required in config/boot.rb
gem 'bootsnap', require: false

# Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
# gem "image_processing", "~> 1.2"

# Use Rack CORS for handling Cross-Origin Resource Sharing (CORS), making cross-origin AJAX possible
# gem "rack-cors"

gem 'activerecord-postgis-adapter'
gem 'rgeo'
gem 'rgeo-proj4'

group :development, :test do
  # See https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
  gem 'debug', platforms: %i[mri mingw x64_mingw]
end

group :development do
  # Speed up commands on slow machines / big apps [https://github.com/rails/spring]
  # gem "spring"
end
```
The tests are mainly nonsense but I wanted to test writing and reading spatial data.

Unfortunately I'm still not able to replicate the issue with this. Are you able to reproduce the issue in a minimal rails app? Or even if you modify your local proj4 gem to see what data is being passed into the `_create` cmethod that could be helpful because I'm having issues reproducing this.

**Comment by BuonOmo** [comment]
@Pe-co do you think you could share an example app link that reproduces the error for you ? That would help us a lot to fix the error.

**Comment by Pe-co** [comment]
I sunk a few hours into it today, and managed to get a repro in a separate project, but its not reproducing stably. I still need to spend some more time minimizing to see what is actually needed.

**Comment by BuonOmo** [comment]
I am reproducing locally :tada:

I'll be offline the next few days, but I (hopefully) downloaded everything to try and find this segfault!

**Comment by keithdoggett** [comment]
Thanks for following up on this @Pe-co this is very helpful

**Comment by BuonOmo** [comment]
@Pe-co @keithdoggett Here is a smaller set of the reproduction: https://github.com/rgeo/rgeo-proj4-segv-issue-39. Basically we should use the thread-safe proj api. I won't have time to open the PR myself soon, so if there is any emergency feel free to take it from here.

**Comment by keithdoggett** [comment]
@BuonOmo I should be able to look into it now that you've found a more reproducible version of it. I was still having issues reproducing on my computer consistently, but your repo looks good. Thanks for the working digging into that.

**Comment by TamerShlash** [comment]
Hello 👋 has this been fixed? still getting Segmentation fault here. Running on Ruby 3.3.0 and `rgeo-proj4-4.0.0`.

```
/path/to/ruby/3.3.0/lib/ruby/gems/3.3.0/gems/rgeo-proj4-4.0.0/lib/rgeo/coord_sys/proj4.rb:249: [BUG] Segmentation fault at 0x0000000105670ad8
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [arm64-darwin23]
```

**Comment by TamerShlash** [comment]
Hi @keithdoggett !

I saw the attempted fix PR as well and have a rough understanding of the issue (project context is being accessed by forked threads if I'm not mistaken). I also had an attempt at fixing that locally but without much success.

I don't have much experience with C extensions but am happy to help, can you explain what is the scope of the change? Would it help taking a look at the way [proj4rb](https://github.com/cfis/proj4rb) handles contexts? I know they have a different use case.

**Comment by BuonOmo** [comment]
Hi @TamerShlash,

Thanks for digging up the topic. I indeed found a fix, but it was only for another issue I found along the way... I don't have a clear remember about this issue, nor the time right no to handle it, but I'll try to help as much as I can.

About the _scope of the change_. It might be related to the proj context: I changed it not to used the default one, and I tried forking (threads and processes) to see if it was still failing, and it worked for my use case. So my guesse would be that there is a specific weird case of forking that requires creating a new context. I think it would be important to first find that case, to be sure we're doing right, and have a test to prove it! Then the scope might be a bit large, but it wouldn't scare that much as it might be a lot of copy-pasta, and only initialization changes and fork hooks (see https://github.com/rails/rails/blob/main/activesupport/lib/active_support/fork_tracker.rb for a reference). I think the hardest is the reproduction, and if you have it and still do not feel confident about the fix, we could pair on this as I think it is at most a 3h work.

However, it may also not be related to context, we really need a reproduction.

About the proj4rb gem, I think you had a good intuition. The [context.rb file](https://github.com/cfis/proj4rb/blob/master/lib/proj/context.rb#L12) is the one we are interested in I think. And their solution is basically to have one thread one context, and an object finalizer that ensures no leak with the proj contexts created

**Comment by hockleyd** [comment]
Just wanted to add a bit more information, although I don't think it does changes the foundation you have outlined, @BuonOmo 

It is a periodic issue for me, in that it doesn't always occur on every test run.  And it is also happening on ruby 3.3.4 and rails 8.0.1, rgeo-proj4, with the segfault happening in the proj4:

/Users/hockleyd/.local/share/mise/installs/ruby/3.3.4/lib/ruby/gems/3.3.0/gems/rgeo-proj4-4.0.0/lib/rgeo/coord_sys/proj4.rb:249: [BUG] Segmentation fault at 0x0000000102ec4a7a

all the other conditions that have been reported are the same.  (This is on arm64 Mac OS Sequoia)

The only other thing is the C backtrace from the ractor dispatch perhaps, but without the debugging symbol table in place, it's not very informative either except at a very general level. 

-- Threading information ---------------------------------------------------
Total ractor count: 1
Ruby thread count for this ractor: 4

-- Machine register context ------------------------------------------------
  x0: 0x00000001ea1fc8ec  x1: 0x0000000000000002  x2: 0x000000018df56cdc
  x3: 0x000000016d02193a  x4: 0x0000000000000008  x5: 0x000000016d021c90
  x6: 0x0000000000000000  x7: 0x0000000000000b00 x18: 0x0000000000000000
 x19: 0x0000600002a3c100 x20: 0x0000600002a3c100 x21: 0x0000000102ec4a7c
 x22: 0x00000001ea1fe000 x23: 0x0000000132134800 x24: 0x000000000000002d
 x25: 0x00000001e808b000 x26: 0x00000001e808b000 x27: 0x00000001ea260000
 x28: 0x00000001e7f30968  lr: 0x00000001805db6b8  fp: 0x000000016d0218c0
  sp: 0x000000016d021880

-- C level backtrace information -------------------------------------------
/Users/hockleyd/.local/share/mise/installs/ruby/3.3.4/lib/libruby.3.3.dylib(rb_vm_bugreport+0xb4c) [0x103b5d658]
/Users/hockleyd/.local/share/mise/installs/ruby/3.3.4/lib/libruby.3.3.dylib(rb_bug_for_fatal_signal+0x100) [0x10399e628]
/Users/hockleyd/.local/share/mise/installs/ruby/3.3.4/lib/libruby.3.3.dylib(sig_do_nothing+0x0) [0x103ac3084]
/usr/lib/system/libsystem_platform.dylib(_sigtramp+0x38) [0x1808c2de4]
/usr/lib/system/libsystem_trace.dylib(_os_log_preferences_refresh+0x24) [0x1805db6b8]
....

**Comment by BuonOmo** [comment]
@hockleyd I see that the segv happens in your teardown. You might be able to isolate something?

```
/Users/hockleyd/.local/share/mise/installs/ruby/3.3.4/lib/libruby.3.3.dylib(_rb_ec_teardown+0xfc) [0x1039a9768]
/Users/hockleyd/.local/share/mise/installs/ruby/3.3.4/lib/libruby.3.3.dylib(_rb_ec_cleanup+0x118) [0x1039a98dc]
```

Besides that I don't think I'd have time to work on it before having a consistant example (even if within a large private codebase)

**Comment by hockleyd** [comment]
I will look more into it, but that won't happen in the next week or two I expect.  As I've not built the system from source at the ruby level itself, it's a fair bit of setup is what I'm expecting as I haven't done C development in quite a while.  I'll see if I can delve further into how I would set up to debug into it.

**Comment by BuonOmo** [comment]
@hockleyd I could do the C dev part without problems, I just need a failure every time I run the tests! And don't worry about deadlines, the sooner the better but no one has report the bug in production, I think the case is rare enough so that it is not too impactfull

**Comment by hockleyd** [comment]
Ok.  I'm not getting a consistently repeatable failure either.  I'll see if I can reduce the test cases and get something cleaner on the rails side to start with at least.  And yes, generally right now I can just break out and rerun the test case (a crash will still allow the tests to keep running, but will not terminate, probably waiting on a join), and get a clean run.   It's definitely not killing and production environment on my end either.
